django/contrib/auth/backends.py

```diff
@@ -97,6 +98,36 @@ def has_module_perms(self, user_obj, app_label):
             for perm in self.get_all_permissions(user_obj)
         )
 
+    def with_perm(self, perm, is_active=True, is_superuser=True, obj=None):
+        UserModel = get_user_model()
+        if not isinstance(perm, (str, Permission)):
+            raise TypeError('The `perm` argument must be a string or a permission instance.')
+        elif isinstance(perm, str):
+            try:
+                app_label, codename = perm.split('.')
+            except ValueError:
+                raise ValueError("Permission name should be in the form 'app_label.perm_name'.") from None
+
+        if obj is not None:
+            return UserModel._default_manager.none()
+
+        if isinstance(perm, Permission):
+            return UserModel._default_manager.filter(
+                Q(is_active=is_active) & (
+                    Q(is_superuser=is_superuser) |
+                    Q(user_permissions=perm) |
+                    Q(groups__permissions=perm)
+                )
+            ).distinct()
+
+        user_q = Q(user_permissions__codename=codename,
+                   user_permissions__content_type__app_label=app_label)
+        group_q = Q(groups__permissions__codename=codename,
+                    groups__permissions__content_type__app_label=app_label)
+        has_permission_q = Q(is_active=is_active) & (Q(is_superuser=is_superuser) | user_q | group_q)
+
+        return UserModel._default_manager.filter(has_permission_q).distinct()
```

Comment:
I would make it more DRY and readable, e.g.
```python
         if isinstance(perm, Permission):
            has_permission = Q(user_permissions=perm) | Q(groups__permissions=perm)
        else:
            has_permission = Q(
                Q(user_permissions__codename=codename) &
                Q(user_permissions__content_type__app_label=app_label)
            ) | Q (
                Q(groups__permissions__codename=codename) &
                Q(groups__permissions__content_type__app_label=app_label)
            )
        return UserModel._default_manager.filter(
            Q(is_active=is_active) &
            Q(Q(is_superuser=is_superuser) | has_permission)
        ).distinct()
```
What do you think?

I wonder if we can use `Exists()` to avoid `distinct()` call.

Reply:
Yes, your suggestion definitely looks better. I'll update the PR. Thanks!

Do you mean the `queryset.exists()` method or does `Exists()` mean something else?

Reply:
I was talking about [Exists() subquery](https://docs.djangoproject.com/en/stable/ref/models/expressions/#exists-subqueries).